[GTK][WPE] Add GWeakPtr to WTF and use it instead of explicit g_object_add|remove_weak_pointer

carlosgcampos opened the pull request:
#### c995526f428610fc057ab7c141f31cbe99a1e2cb
<pre>
[GTK][WPE] Add GWeakPtr to WTF and use it instead of explicit g_object_add|remove_weak_pointer
<a href="https://bugs.webkit.org/show_bug.cgi?id=251258">https://bugs.webkit.org/show_bug.cgi?id=251258</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

It&apos;s very common to forget to remove a weak references, GWeakPtr ensures
weak references are always removed and makes easier to create weak
GObjects.

* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/GWeakPtr.h: Added.
(WTF::GWeakPtr::GWeakPtr):
(WTF::GWeakPtr::~GWeakPtr):
(WTF::GWeakPtr::get const):
(WTF::GWeakPtr::reset):
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(webkitDownloadCreate):
(webkit_download_get_web_view):
(_WebKitDownloadPrivate::~_WebKitDownloadPrivate): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetLoadDecisionForIcon):
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
(UIClient::~UIClient): Deleted.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/259482@main">https://commits.webkit.org/259482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3d2ecf9281e939c960ea5eded8627c02a4d2a8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105025 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114288 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174471 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108928 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5026 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113301 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110781 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94871 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7442 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92910 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5172 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7536 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30276 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13591 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101606 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9325 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25349 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3483 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/glib/GWeakPtr.h:
```diff
@@ -0,0 +1,111 @@
+    GWeakPtr(GWeakPtr&& other)
+    {
+        auto* ptr = other.get();
+        reset(ptr);
```
I would have used a single line here, but no strong preference:

```suggestion
        reset(other.get());
```

mcatanzaro replied:
Oh yeah, I noticed that too but just forgot to point it out. :P

aperezdc reviewed (approved):
Nice, one has to like how the new `GWeakPtr` template allows simplifying code elsewhere 🤩

mcatanzaro reviewed (approved):
Nice. Much safer than before.

I see the web inspector code was never removing its weak pointers, so this should fix memory corruption there, right? It must so, unless the GtkWidgets are guaranteed to outlive the WebInspectorUIProxy, but if that were so you wouldn't need the weak pointers in the first place. And this was in the UI process, too. :/

carlosgcampos committed:
[GTK][WPE] Add GWeakPtr to WTF and use it instead of explicit g_object_add|remove_weak_pointer
https://bugs.webkit.org/show_bug.cgi?id=251258

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

It's very common to forget to remove a weak references, GWeakPtr ensures
weak references are always removed and makes easier to create weak
GObjects.

* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/GWeakPtr.h: Added.
(WTF::GWeakPtr::GWeakPtr):
(WTF::GWeakPtr::~GWeakPtr):
(WTF::GWeakPtr::get const):
(WTF::GWeakPtr::reset):
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(webkitDownloadCreate):
(webkit_download_get_web_view):
(_WebKitDownloadPrivate::~_WebKitDownloadPrivate): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetLoadDecisionForIcon):
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
(UIClient::~UIClient): Deleted.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/259482@main
